README.md
birdback — simple fast Ubuntu backups
=====================================

birdback is a simple, fast and open-source backup tool for Ubuntu Linux users:
 - an icon is shown in the global menu bar of Ubuntu for controlling the application
 - USBs/HDDs are detected and mounted automatically for backup
 - backups are made of the user's home folder **ONLY** (for now)
 - backups are simply files being copied over in an efficient manner (see section below for more info)
 - when it comes time for the backup to be restored, the user can choose how and what files they copy back to the main HDD
 
What birdback does not do:
 - cross-platform backups (Ubuntu Linux only)
 - versioned backups
 - encryption

Licensed under GPLv3 to Liam Edwards-Playne and contributors.

## Why is birdback fast?
Even though birdback is written in higher-level langauge such as Python, it is still pretty fast. This is because of the very simple two-pass backup process.

When birdback is instructed to backup the home folder, it walks the directory tree and constructs a list of files to backup based on whether they are already on the backup destination and are the newest version of the file. This is inherently fast because it is just read operations. It is made even faster by using scandir, a quicker alternative to the standard lib's `os.walk`. 
The second stage is copying these files to the backup destination, which is just writing data.

## Install
I'm going to make this more automated later, perhaps in a Debian package.

 1. Clone git repo.
 2. `sudo apt-get install python3 python3-dev python3-setuptools python3-gobject`
 3. `sudo python setup.py build`
 4. `sudo python setup.py install`


HISTORY.md
0.1.0 (2013-12-5)
=================

First major release.


src/birdback/test.py
from gi.repository import Gtk
from gi.repository import AppIndicator3 as appindicator
import os
import signal

class BirdbackGUI(object):
	def __init__(self):
		self.pidfile = os.path.expanduser("~/.birdback.pid")
		self.check_pid()
		try:
			self.indicator = Indicator()
		except Exception as e:
			print(e)
			print("Critical error. Exiting.")
			self.exit(1)

	def signal_exit(self, signum, frame):
		print('Recieved signal: ', signum)
		print('Quitting...')
		self.exit()
	
	def exit(self, code=0):
		try:
			print('stopping')
			#self.xflux_controller.stop()
		except MethodUnavailableError:
			pass
		os.unlink(self.pidfile)
		Gtk.main_quit()
		sys.exit(code)

	def check_pid(self):
		pid = os.getpid()

		running = False # Innocent...
		if os.path.isfile(self.pidfile):
			try:
				oldpid = int(open(self.pidfile).readline().rstrip())
				try:
					os.kill(oldpid, 0)
					running = True # ...until proven guilty
				except OSError as err:
					if err.errno == errno.ESRCH:
						# OSError: [Errno 3] No such process
						print("stale pidfile, old pid: ", oldpid)
			except ValueError:
				# Corrupt pidfile, empty or not an int on first line
				pass
		if running:
			print("birdback is already running, exiting")
			sys.exit()
		else:
			file(self.pidfile, 'w').write("%d\n" % pid)
	
	def __del__(self):
		self.exit()
	
	def run(self):
		Gtk.main() # run Gtk run!

class Indicator(object):
	def __init__(self):
		# See http://developer.ubuntu.com/api/devel/ubuntu-13.10/python/AppIndicator3-0.1.html
		self.indicator = appindicator.Indicator(
		  "birdback-indicator",
		  "birdback",
		  appindicator.CATEGORY_APPLICATION_STATUS)

		self.setup_indicator()

	def setup_indicator(self):
		self.indicator.set_status(appindicator.STATUS_ACTIVE)
		self.indicator.set_icon('birdback')
		self.indicator.set_menu(self.create_menu())

	def create_menu(self):
		menu = Gtk.Menu()

		self.add_menu_item("_Pause f.lux", self._toggle_pause,
				menu, MenuItem=Gtk.CheckMenuItem)
		self.add_menu_item("_Preferences", self._open_preferences, menu)
		self.add_menu_separator(menu)
		self.add_menu_item("Quit", self._quit, menu)

		return menu

	def add_menu_item(self, label, handler, menu, event="activate", MenuItem=Gtk.MenuItem, show=True):
		item = MenuItem(label)
		item.connect(event, handler)
		menu.append(item)
		if show:
			item.show()
		return item

	def add_menu_separator(self, menu, show=True):
		item = Gtk.SeparatorMenuItem()
		menu.append(item)
		if show:
			item.show()

	def _toggle_pause(self, item):
		self.xflux_controller.toggle_pause()

	def _open_preferences(self, item):
		self.fluxgui.open_preferences()

	def _quit(self, item):
		self.fluxgui.exit()

if __name__ == '__main__':
	try:
		app = BirdbackGUI()
		signal.signal(signal.SIGTERM, app.signal_exit)
		app.run()
	except KeyboardInterrupt:
		app.exit()


src/birdback/controller.py
import view
import model

import scandir
import time
import os
import os.path
import sys
import shelve
import getpass
import subprocess
import glob
import shutil
import errno

import logging

import pyinotify
from gi.repository import Gtk
from gi.repository import GObject

GObject.threads_init()

class Controller(object):
	def __init__(self):
		self.backupMediums = {}
		self.pidfile = os.path.expanduser("~/.birdback.pid")
		
		# Check that birdback isn't already running
		# -----------------------------------------
		pid = os.getpid()
		
		running = False # Innocent...
		if os.path.isfile(self.pidfile):
			try:
				oldpid = int(open(self.pidfile).readline().rstrip())
				try:
					os.kill(oldpid, 0)
					running = True # ...until proven guilty
				except OSError as err:
					if err.errno == os.errno.ESRCH:
						# OSError: [Errno 3] No such process
						print("stale pidfile, old pid: ", oldpid)
			except ValueError:
				# Corrupt pidfile, empty or not an int on first line
				pass
		if running:
			print("birdback is already running, exiting")
			sys.exit()
		else:
			open(self.pidfile, 'w').write("%d\n" % pid)
			print("pidfile updated")
			
		# Load preferences from disk
		# --------------------------
		preferencePath = os.path.join(os.path.expanduser("~"), ".local", "share", 'birdback')
		self.preferences = shelve.open(preferencePath)
		print("Preferences loaded")
		
		# Instantiate file system watchers
		# --------------------------------
		watchManager = pyinotify.WatchManager()
		
		class BackupMediaDetector(pyinotify.ProcessEvent):
			def __init__(self, controller):
				self.controller = controller
			
			def process_IN_CREATE(self, event):
				path = event.pathname
				if path.startswith('/dev/disk/by-id/usb'):
					# Try to automatically mount it
					devicePath = os.path.realpath("/dev/disk/by-id/"+os.readlink(path))
					try:
						# We use udisksctl here because it mounts to /media/USERNAME/XXX instead of simply /media/XXX
						subprocess.check_output(['udisksctl', 'mount', '--block-device', devicePath])
					except:
						print('Error while mounting path: '+path)
					return
				else:
					print('HDD/USB inserted at: ' + path)
					if path not in self.controller.backupMediums:
						self.controller.backupMediums[path] = model.BackupMedium(path)
					self.controller.view.drive_inserted(self.controller.backupMediums[path])

			def process_IN_DELETE(self, event):
				path = event.pathname
				print('HDD/USB removed at: ' + path)
				if path in self.controller.backupMediums:
					self.controller.view.drive_removed(self.controller.backupMediums[path])
					del self.controller.backupMediums[path]
		
		self.backupMediaWatcher = pyinotify.ThreadedNotifier(watchManager, BackupMediaDetector(self))
		self.backupMediaWatcher.start()
		
		watchManager.add_watch(os.path.join('/media', getpass.getuser()), pyinotify.IN_DELETE | pyinotify.IN_CREATE, rec=False)
		watchManager.add_watch('/dev/disk/by-id/', pyinotify.IN_CREATE, rec=False)
		print("Added watch for USB/HDDs")
	
	def run(self):		
		# Instantiate the view (menu bar)
		# -------------------------------
		self.view = view.View(self)
		print("View instantiated")
		
		# Detect existing HDDs/USBs
		# -------------------------
		os.chdir("/dev/disk/by-id")
		try:
			for path in glob.glob("usb*"):
				# Try to automatically mount it
				devicePath = os.path.realpath("/dev/disk/by-id/"+os.readlink(path))
				mounts = open("/proc/mounts")
				for line in mounts:
					parts = line.split(' ')
					if parts[0] == devicePath and parts[1].startswith(os.path.join('/media', getpass.getuser())):
						path = parts[1]
						self.backupMediums[path] = model.BackupMedium(path)
						self.view.drive_inserted(self.backupMediums[path])
				mounts.close()
		except:
			print('Error while detecting existing HDDs/USBs: '+path)
	
		# Start doing stuff
		# -----------------
		print("Running main loop")
		Gtk.main()
	
	def signal_exit(self, _0, _1):
		self.quit()
	
	def quit(self, code=0):
		print("Quitting...")
		try:
			self.backupMediaWatcher.stop()
			Gtk.main_quit()
			self.preferences.close()
		except Exception as e:
			print("Exception while quitting:")
			print(e)
		
		# DO NOT CHANGE THE ORDER OF CALLS BELOW
		try:
			os.remove(self.pidfile)
		except OSError:
			pass
	
	def backup(self, backupMedium, progress_callback):
		filesToBackup = []
		
		# progress_callback("x/4 backing up installed programs/packages")
		# dpkg --get-selections > ~/Package.list
		# cp /etc/apt/sources.list ~/sources.list
		# apt-key exportall > ~/Repo.keys
		
		progress_callback("1/3 deleting old files from backup")
		self.deleteOldFiles(backupMedium, progress_callback)
		
		# If backup device was removed while old files were being deleted in deleteOldFiles, then the directory walk just exits, which is why I test here
		if not os.path.exists(backupMedium.path):
			raise Exception("Backup device was removed")
		
		progress_callback("2/3 scanning documents")
		filesToBackup.extend(self.home_files_to_backup(backupMedium, progress_callback))
				
		progress_callback("3/3 backing up")
		for i, src_file in enumerate(filesToBackup):
			progress = float(i / len(filesToBackup))
			progress_callback("3/3 backing up ({0:.1f}%)".format(100*progress), log=False)
			
			dst_dir = os.path.join(backupMedium.path, os.path.dirname(src_file[1:]))
			try:
				os.makedirs(dst_dir, exist_ok=True)
				shutil.copy2(src_file, dst_dir, follow_symlinks=False)
			except OSError as e:
				if e.errno == errno.ENXIO:
					# XXX I have no friggin' idea what this is or means
					# Pops up sometimes for GNOME temp files in .local, I don't really care
					print("ENXIO for "+src_file)
					continue
			except Exception as e:
				if not os.path.exists(backupMedium.path):
					raise Exception("Backup device was removed")
				elif not os.path.exists(src_file):
					print("[{0}]: not backing up because file doesn't exist - {1}".format(backupMedium.name, src_file))
					continue
				else:
					# Something else failed like:
					#  - backup medium has no space left
					#  - we couldn't create the directory on the backup medium
					#  - we couldn't copy the file over
					raise e
		
		progress_callback("backup complete")
		
	
	def home_files_to_backup(self, backupMedium, progress_callback):
		BACKUP_PATH = os.path.expanduser("~")
		EXCLUDES = [
			'.cache',
			'.ccache',
			'Downloads',
			'tmp'
		]
		
		filesToBackup = []
		filesProcessed = 0
		
		for root, dirs, files in scandir.walk(BACKUP_PATH, topdown=True):
			if not os.path.exists(backupMedium.path):
				raise Exception("Backup device was removed")
			
			# Common excludes
			if root == BACKUP_PATH:
				dirs[:] = [d for d in dirs if d not in EXCLUDES]
			
			# Special excludes
			if root == (os.path.join(BACKUP_PATH, '.local/share')):
				if 'Trash' in dirs: dirs.remove('Trash')
			
			if not os.path.exists(os.path.join(backupMedium.path, root[1:])):
				for f in files:
					filesProcessed += 1
					progress_callback("2/3 scanning documents ({0} processed)".format(filesProcessed), log=False)
					filesToBackup.append(os.path.join(root, f))
			else:
				for f in files:
					filesProcessed += 1
					progress_callback("2/3 scanning documents ({0} processed)".format(filesProcessed), log=False)
										
					absolute_file = os.path.join(root, f)
					remote_mtime = -1
					try: 
						remote_mtime = os.path.getmtime(os.path.join(backupMedium.path, absolute_file[1:]))
					except:
						pass
					try:
						if os.path.getmtime(absolute_file) > remote_mtime:
							filesToBackup.append(absolute_file)
					except OSError as e:
						pass
			
		return filesToBackup
	
	def deleteOldFiles(self, backupMedium, progress_callback):
		filesProcessed = 0		
		
		PATH = os.path.expanduser("~")
		for root, dirs, files in scandir.walk(os.path.join(backupMedium.path, PATH[1:]), topdown=True):
			for d in dirs:
				absolute_path = os.path.join(root, d)
				if not os.path.exists('/'+os.path.relpath(absolute_path,backupMedium.path)):
					try:
						shutil.rmtree(absolute_path)
					except:
						pass
			
			for f in files:
				filesProcessed += 1
				progress_callback("1/3 deleting old files from backup ({0} processed)".format(filesProcessed), log=False)
				
				absolute_file = os.path.join(root, f)
				
				if not os.path.exists('/'+os.path.relpath(absolute_file, backupMedium.path)):
					try:
						os.remove(absolute_file)
					except:
						pass










src/birdback/view.py
from gi.repository import Gtk
from gi.repository import GObject
from gi.repository import AppIndicator3 as appindicator
from gi.repository import Notify
import logging

class View(object):
	QUIT_ICON = '/usr/share/icons/Humanity/actions/128/process-stop.svg'
	ACTIVE_ICON = '/usr/local/share/icons/hicolor/scalable/apps/birdback.svg'
	ATTENTION_ICON = '/usr/local/share/icons/hicolor/scalable/apps/birdback-active.svg'
	DO_BACKUP_ICON = '/usr/share/icons/Humanity/actions/48/forward.svg'
	
	def __init__(self, controller):
		self.controller = controller
		# See http://developer.ubuntu.com/api/devel/ubuntu-13.10/python/AppIndicator3-0.1.html
		self.indicator = appindicator.Indicator.new(
		  "birdback-indicator",
		  "birdback",
		  appindicator.IndicatorCategory.APPLICATION_STATUS)
		self.indicator.set_status(appindicator.IndicatorStatus.ACTIVE)
		self.indicator.set_icon(View.ACTIVE_ICON)
		self.indicator.set_attention_icon(View.ATTENTION_ICON)
		self.indicate_bliss()
		self.menu = self.create_menu()
		self.indicator.set_menu(self.menu)
		Notify.init('birdback')
		self.backupControls = {}
	
	def indicate_activity(self):
		self.indicator.set_status(appindicator.IndicatorStatus.ATTENTION)
	
	def indicate_bliss(self):
		self.indicator.set_status(appindicator.IndicatorStatus.ACTIVE)
	
	def add_menu_item(self, label, handler, event="activate", MenuItem=Gtk.MenuItem, show=True):
		item = MenuItem(label)
		item.connect(event, handler)
		self.menu.append(item)
		if show:
			item.show()
		return item
	
	def add_image_menu_item(self, label, handler, imgPath, event="activate", show=True):
		img = Gtk.Image()
		img.set_from_file(imgPath)
		item = Gtk.ImageMenuItem(label)
		item.set_image(img)
		item.connect(event, handler)
		self.menu.append(item)
		item.show()
		return item
	
	def create_menu(self):
		self.menu = Gtk.Menu()
		self.menu.show_all()
		
		self.add_image_menu_item('Quit', self.quit, View.QUIT_ICON)
		
		return self.menu
	
	# I don't know why I have to do this. But it works.
	def update_view(self):
		while Gtk.events_pending():
			Gtk.main_iteration_do(False)
	
	def show_notification(self, message):
		n = Notify.Notification.new('BirdBack', message, None)
		n.show()
	
	def drive_inserted(self, backupMedium):		
		def default_backup_label():
			return "Backup "+backupMedium.name
		def backup(x):
			print("Starting backup of " + backupMedium.path)
			menuItem = self.backupControls[backupMedium.path]
			menuItem.set_sensitive(False)
			self.indicate_activity()
			self.update_view()
			
			def progress_callback(progress, log=True):
				menuItem.set_label("Backing up {0} \n{1}".format(backupMedium.name, progress))
				if log:
					print("backup "+backupMedium.name+": "+progress)
				self.update_view()
			try:
				self.controller.backup(backupMedium, progress_callback)
				self.show_notification("Backup complete for "+backupMedium.name)
			except Exception as err:
				notice = "Backup of {0} failed: {1}".format(backupMedium.name, err)
				print(notice)
				logging.exception('')
				self.show_notification(notice)
			finally:
				menuItem.set_sensitive(True)
				menuItem.set_label(default_backup_label())
				self.indicate_bliss()
				self.update_view()
			
		self.backupControls[backupMedium.path] = self.add_image_menu_item(default_backup_label(), backup, View.DO_BACKUP_ICON)
		self.backupControls[backupMedium.path].sensitive = False
	
	def drive_removed(self, backupMedium):
		if backupMedium.path in self.backupControls:
			self.menu.remove(self.backupControls[backupMedium.path])
	
	def quit(self, _):
		Notify.uninit()
		self.controller.quit()
